### ISSUE TITLE
Implement interleaving AES-CBC-HMAC-SHA in ENCRYPT-then-MAC order

### ISSUE THREAD

**Issue by zorrorffm** [openssl/openssl#19932]
AES-CBC-HMAC-SHA provides authenticated encryption and is widely used in real security world.  It is one of the ciphersuites in TLS 1.2 and it is also one of the cipher-authentication features in DPDK OpenSSL PMD(https://doc.dpdk.org/guides/cryptodevs/openssl.html). The operation order with AES-CBC and HMAC depends on the use case, for TLS 1.2 the operation order is MAC-then-ENCRYPT while for DPDK OpenSSL PMD, the operation order is ENCRYPT-then-MAC. 

OpenSSL is capable of running AES-CBC-HMAC-SHA in either MAC-then-ENCRYPT or ENCRYPT-then-MAC order because OpenSSL implements AES-CBC and SHA respectively, user can combine them at will. But from performance perspective,  interleaving AES-CBC with HMAC-SHA should be the better implementation, we had perf test on non-interleaving AES-CBC-HMAC-SHA and interleaving AES-CBC-HMAC-SHA in DPDK, the result is following

| Byte   |  MOps(interleaving AES-CBC-HMAC-SHA1)   |   MOps(non-interleaving AES-CBC-HMAC-SHA1) |
| -------| ------------------------------------------------- | ----------------------------------------------------|
|   64    |  6.0440                                                             |   0.5154                                                               |
|  128   |  4.3194                                                             |   0.5034                                                               |
|  256   |  2.7646                                                             |   0.4706                                                               |
|  512   |  1.6436                                                             |   0.4011                                                               |
| 1024  |  0.9072                                                             |   0.3421                                                               |
| 2048  |  0.4774                                                             |   0.2518                                                               |
| 4096  |  0.2459                                                             |   0.1616                                                               |

The interleaving AES-CBC-HMAC-SHA1 implementation is from Armv8 PMD(https://doc.dpdk.org/guides/cryptodevs/armv8.html), the non-interleaving AES-CBC-HMAC-SHA1 implementation is from OpenSSL PMD(https://doc.dpdk.org/guides/cryptodevs/openssl.html).  This result shows interleaving implementation outperforms non-interleaving implementation a lot.

OpenSSL provides interleaving AES-CBC-HMAC-SHA in e_aes_cbc_hmac_sha1.c and e_aes_cbc_hmac_sha256.c, but this implementation is for internal TLS protocol, it's not well wrapped by EVP API, and it's only MAC-then-ENCRYPT implementation.

So let's summarize the situation, there are needs for AES-CBC-HMAC-SHA in both MAC-then-ENCRYPT and ENCRYPT-then-MAC order, interleaving AES-CBC with HMAC-SHA has better performance. OpenSSL only provides interleaving AES-CBC-HMAC-SHA in MAC-then-ENCRYPT order.

**Here comes my proposal**

Make AES-CBC-HMAC-SHA in OpenSSL capable of running ENCRYPT-then-MAC in interleaving way, and the change won't bring any side effect to current existing MAC-then-ENCRYPT implementation.  To achieve this, we can do

1.  Add the interface to allow user chose MAC-then-ENCRYPT or ENCRYPT-then-MAC
    - Add an integer member named enc_then_mac in structure prov_aes_hmac_sha_ctx_st,  this member is used to indicate the order of authentication and cipher.
    - Define OSSL_CIPHER_PARAM_ENC_THEN_MAC in core_names.h, it provide user the access to set enc_then_mac above.
    - Add implementation in function aes_set_ctx_params to set enc_then_mac
 2. Implement interleaving AES-CBC-HMAC-SHA in ENCRYPT-then-MAC order, this implementation is assembly code and architecture related.
 3. Provide fallback implementation(pure C code) for those architectures not providing interleaving implementation.

**Comment by bernd-edlinger** [comment]
SHA1 is as deprecated as it can be...

**Comment by tom-cosgrove-arm** [comment]
Agreed, but it is still considered safe in HMAC, and used in existing protocols (hence the desire to implement an optimised ENCRYPT-then-MAC with AES-CBC+HMAC-SHA1)

**Comment by bernd-edlinger** [comment]
And there is a reason why ENCRYPT-then-MAC is not considered safe either.

**Comment by tom-cosgrove-arm** [comment]
Huh? [Encrypt-then-MAC is generally considered the best approach](https://crypto.stackexchange.com/questions/202/should-we-mac-then-encrypt-or-encrypt-then-mac) - it pretty much gives you authenticated encryption (integrity of ciphertext).

Yes, we prefer "rolled up" AEAD modes these days, since there are bad ways of combining non-authenticated encryption with MACs; this issue is us asking if you would accept code that does just such a "rolled up" AES-CBC+HMAC-SHA1 (which doesn't have such encryption-mode-interferes-with-MAC issues).

**Comment by slontis** [comment]
> Agreed, but it is still considered safe in HMAC, and used in existing protocols (hence the desire to implement an optimised ENCRYPT-then-MAC with AES-CBC+HMAC-SHA1)

See https://csrc.nist.gov/news/2022/nist-transitioning-away-from-sha-1-for-all-apps

**Comment by zorrorffm** [comment]
SHA1 is not secure any more, but as https://en.wikipedia.org/wiki/SHA-1 said,  SHA1 is still secure for HMAC.

**Comment by tom-cosgrove-arm** [comment]
We would like (indeed would prefer) to do the SHA-256 variant, and that would be next. However, as the NIST document says

> The transition will be completed by December 31, 2030

and even if it's just for the next 3 of those 8 years before the transition is complete, having systems using a better-optimised version is a lot of energy saved.

**Comment by t8m** [comment]
IMO unless the patch regresses performance or security properties of other modes (in particularly the M-t-E) or on other platforms, it would be welcome.

Of course this would be for master branch only.

**Comment by t8m** [comment]
It was implemented at least for ARM.
